src/huggingface_hub/utils/_verification.py

```diff
@@ -0,0 +1,204 @@
+    )
+
+
+def resolve_expected_hash(entry: object) -> tuple[HashAlgo, str]:
+    """
+    Return the algorithm and expected hash for a remote entry.
+    Prefers LFS sha256 if available; falls back to git blob_id (sha1).
+    """
+    lfs = getattr(entry, "lfs", None)
+    lfs_sha = getattr(lfs, "sha256", None) if lfs is not None else None
+    if lfs_sha is None and isinstance(lfs, dict):
+        lfs_sha = lfs.get("sha256")
+
+    if lfs_sha:
+        return ("sha256", str(lfs_sha).lower())
+
+    blob_id = getattr(entry, "blob_id", None)
+    if blob_id:
+        return ("git-sha1", str(blob_id).lower())
+
+    raise ValueError("Remote entry missing checksum (no blob_id or lfs.sha256)")
+
+
+def compute_file_hash(path: Path, algorithm: HashAlgo, *, git_hash_cache: dict[Path, str]) -> str:
+    """
+    Compute the checksum of a local file using the requested algorithm.
+    """
+
+    def _sha256(p: Path, cache: dict[Path, str]) -> str:
+        with p.open("rb") as stream:
+            return sha_fileobj(stream).hex()
+
+    def _git_sha1(p: Path, cache: dict[Path, str]) -> str:
+        try:
+            return cache[p]
+        except KeyError:
+            with p.open("rb") as stream:
+                data = stream.read()
+            digest = git_hash(data)
+            cache[p] = digest
+            return digest
+
+    HASHERS: dict[HashAlgo, Callable[[Path, dict[Path, str]], str]] = {
+        "sha256": _sha256,
+        "git-sha1": _git_sha1,
+    }
+
+    try:
+        return HASHERS[algorithm](path, git_hash_cache)
+    except KeyError:
+        # Should be unreachable, but keeps type checker happy
+        raise ValueError(f"Unsupported hash algorithm: {algorithm}")
+
+
+def verify_maps(*, remote_by_path: dict[str, object], local_by_path: dict[str, Path], revision: str) -> Verification:
+    """Compare remote entries and local files and return a verification result."""
+    remote_paths = set(remote_by_path)
+    local_paths = set(local_by_path)
+
+    missing = sorted(remote_paths - local_paths)
+    extra = sorted(local_paths - remote_paths)
+    both = sorted(remote_paths & local_paths)
+
+    mismatches: list[Mismatch] = []
+    git_hash_cache: dict[Path, str] = {}
+
+    for rel_path in both:
+        entry = remote_by_path[rel_path]
+        local_path = local_by_path[rel_path]
+
+        try:
+            algorithm, expected = resolve_expected_hash(entry)
+            actual = compute_file_hash(local_path, algorithm, git_hash_cache=git_hash_cache)
+        except OSError as exc:
+            mismatches.append(
+                Mismatch(path=rel_path, expected="<unavailable>", actual=f"io-error:{exc}", algorithm="io")
+            )
+            continue
+        except ValueError as exc:
+            mismatches.append(
+                Mismatch(path=rel_path, expected="<unavailable>", actual=f"meta-error:{exc}", algorithm="meta")
+            )
+            continue
```

Comment:
I don't understand this part :confused: Shouldn't "algorithm" be `git-hash` and `sha256`? Also why could a OSError or ValueError be raised since `compute_file_hash` is not raising anything?